graph/patterns/longRunningOperations.md

```diff
@@ -0,0 +1,192 @@
+presentation makes things simpler for our users and tooling chain.
+
+In general, Microsoft Graph API guidelines for LRO follow [Microsoft REST API
+Guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#13-long-running-operations).
+
+A single deviation from the base guidelines is that Microsoft Graph API standards require that you use the following response headers:
+
+- The Content-Location header indicates the location of a RELO resource.
+  - The API response says the targeted resource is being created by returning a 201 status code and the resource URI is provided in the Content-Location header, but the response indicates that the request is not completed by including "Provisioning" status.
+
+- The Location header indicates the location of a new stepwise operation LRO resource.
+  - The API response says the operation resource is being created at the URL provided in the Location header and indicates that the request is not completed by including a 202 status code.
+
+## When to use this pattern
+
+Any API call that is expected to take longer than 1 second in the 99th percentile should use the long running operations pattern.
+
+How do you select which flavor of LRO pattern to use? An API designer can follow these
+heuristics:
+
+1.  If a service can create a resource with a minimal latency and continue
+    updating its status according to the well-defined and stable state
+    transition model until completion, then the RELO model is the best choice.
+
+2.  Otherwise, a service should follow the Stepwise Operation pattern. 
+ 
+
+## Issues and considerations
+
+- One or more clients MUST be able to monitor and operate on the same resource at the same time.
+
+- The state of the system SHOULD always be discoverable and testable. Clients
+    SHOULD be able to determine the system state even if the operation tracking
+    resource is no longer active. Clients MAY issue a GET on some resource to
+    determine the state of a long running operation.
+
+-  The long running operations pattern SHOULD work for clients looking to "fire and forget"
+    and for clients looking to actively monitor and act upon results.
+
+-  The long running operations pattern might be supplemented by the [change notification
+    pattern](change-notification.md).
+
+-  Cancellation of a long running operation does not explicitly mean a rollback. On a per API-defined case, it
+    might mean a rollback, or compensation, or completion, or partial completion,
+    etc. Following a canceled operation, the API should return a consistent state that allows
+    continued service.
+
+-  A recommended minimum retention time for a stepwise operation is 24 hours.
+    Operations SHOULD transition to "tombstone" for an additional period of time
+    prior to being purged from the system.
+    
+-  Services that provides a new operation resource MUST support GET semantics on the operation.
+
+
```

Comment:
Should we allow the case where an API producer sometimes immediately completes an operation and other times returns an LRO?

Reply:
:-) I think so, but Gareth disagrees.  We should have a longer conversation about this.

Reply:
I think that if API response takes long time for P99 then it should be LRO and edge cases may be treated as errors

Reply:
In my opinion for the resource-based pattern it's OK if sometimes the operation completes immediately and other times it's a LRO as in the case the shape of the API response is the same. For the stepwise pattern, my thinking is if an API can return a LRO then it should always return an LRO (even if the LRO is created in the completed state) that way API consumers don't have to deal with two different shapes of response.

We may even want to consider a metadata annotation to make any API that uses the step-wise LRO pattern

Reply:
It's a good recommendation. Will add to the pattern